### PR TITLE
Move the guest should verify email logic to the order class

### PR DESCRIPTION
**Pull request by hsingyuc**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR move the guest should verify email logic to the order class so we can reuse email is valid check.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a guest pay-for-order order with (ex. WC Pre-Orders)
3. Find the order and update the status to `pending payment`
4. Open the `Customer payment page →` in incognito
5. See the pay-for-order order without email verification because we are in the [grace period](https://github.com/woocommerce/woocommerce/blob/7ec5a88573a49cdb8b432a803007ccf7447fe5ae/plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php#L419)
6. Comment out the grace period or wait till the time pass
7. Go to the pay-for-order page again
8. See the verify email address page.

### Screenshot
Screenshot 2024-01-21 at 3 40 00 PM

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [x] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Move the guest should verify email logic to the user utils

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php
Comment (leonardola) [inline plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php]: This one liner is quite hard to understand at first. Let's take the opportunity and refactor it into a more readable code. Maybe change to something like this:

```
// If we cannot match the order with the current user, ask that they verify their email address.
$supplied_email_matches_order_email = sanitize_email( wp_unslash( filter_input( INPUT_POST, 'email' ) ) ) === $order->get_billing_email();
$nonce_is_valid                     = wp_verify_nonce( filter_input( INPUT_POST, 'check_submission' ), 'wc_verify_email' );
$supplied_email                     = null;

if ( $supplied_email_matches_order_email &&  $nonce_is_valid ) {
	$supplied_email = sanitize_email( wp_unslash( filter_input( INPUT_POST, 'email' ) ) );
}
```

Also it ends up checking if the email matches with the order again in the new function. Do we really need to also check it here?

**Review by leonardola — CHANGES_REQUESTED** [review]
Generally the changes look good. Tests went well. But I left a comment about making the code more readable. 

Also the wc-order class is already huge so I'm a little unease about moving more code into it.

We'll need to check with a team that is more familiar with WooCommerce core code to  also review this PR before it gets merged.

File: plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php
Comment (hsingyuc) [reply plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php]: >This one liner is quite hard to understand at first. Let's take the opportunity and refactor it into a more readable code.

Good idea, thank you! 

>Also it ends up checking if the email matches with the order again in the new function. Do we really need to also check it here?

Thank you for catching that!  Both issues are fixed here [cf90234](https://github.com/woocommerce/woocommerce/pull/43834/commits/cf9023489423497b3cb763b055a186df2ffc9c9a).

**Review by leonardola — APPROVED** [review]
Changes look good now. Thanks for the fixes.

**Comment by hsingyuc** [comment]
Hi @jonathansadowski, can you help us review this PR, we'll need it before Jan 31st for testing.  Thank you!

**Review by jonathansadowski — DISMISSED** [review]
> Hi @jonathansadowski, can you help us review this PR, we'll need it before Jan 31st for testing. Thank you!

Hey @hsingyuc I cycled the tests on this by closing and reopening the PR, as after the changelog was added by our automation, the checks weren't automatically triggered (a fix for this is in the works). I noticed that there was a [lint issue](https://github.com/woocommerce/woocommerce/actions/runs/7699790381/job/20982630002?pr=43834) that should be addressed before this was run.

> We'll need to check with a team that is more familiar with WooCommerce core code to also review this PR before it gets merged.

I'm going to add @woocommerce/proton as a reviewer to this PR for that. @woocommerce/vortex is primarily involved with the tooling surrounding the monorepo.

**Comment by barryhughes** [comment]
> Also the wc-order class is already huge so I'm a little unease about moving more code into it.

Fwiw, I agree with this. Less because of the *size* of the existing `WC_Order` class (though, yep, it's definitely grown pretty big), but because its core job is to model orders, rather than to govern access to a frontend view.

Could this logic live inside `OrderAuthorizationTrait`? Even though that is in the Blocks namespace, I'm sure the shortcode class could still reference it (or, if that is not palatable, a new trait or utility class could be created).

**Comment by hsingyuc** [comment]
>I noticed that there was a [lint issue](https://github.com/woocommerce/woocommerce/actions/runs/7699790381/job/20982630002?pr=43834) that should be addressed before this was run.

Thank you, @jonathansadowski ! Fixed in [eea1892](https://github.com/woocommerce/woocommerce/pull/43834/commits/eea1892dc3af83c9fb064f2918a877156b2c218e).

@barryhughes  This is more of a permission check, it's only checking if the email is valid. I put it inside `class-wc-order` because we put `key_is_valid` there, and they seem like they should be together.  I don't think we want to put it inside `OrderAuthorizationTrait` because that's for endpoints and might be weird for the shortcode to use.  I found `/Utilities/OrderUtil.php` but it doesn't seem to fit well with the other methods in there.  Any suggestions?

**Comment by barryhughes** [comment]
Hi @hsingyuc,

> I found /Utilities/OrderUtil.php but it doesn't seem to fit well with the other methods in there.

*True.* Although, it's described as a spot for any order-related helpers—so it wouldn't be the worst fit (but I take your point). If that's uncomfortable, how about either of:

- `Automattic\WooCommerce\Internal\Utilities\Users` (and renaming the method to `should_user_verify_order_email`).
- Or, placing it in a new type `Automattic\WooCommerce\Internal\Utilities\Checkout`.

Both have the advantage we are not adding frontend validation logic to a data model, and both are in the internal namespace which can be used from both block and traditional shortcode logic, but gives us the freedom to refactor further in the future.

**Comment by hsingyuc** [comment]
Thank you, @barryhughes! Fixed in [bc14f07](https://github.com/woocommerce/woocommerce/pull/43834/commits/bc14f079714ee7f8e825806a2e8e97e3b40e0b25) and moved the function to user utils.

**Comment by barryhughes** [comment]
Closing and re-opening, just to try and get our checks to run (they seem to have unexepectedly stopped *("waiting for status to be reported").*

**Comment by barryhughes** [comment]
Hi @hsingyuc, I wanted to take a bit of time to test some scenarios, because the rollout of the original change (adding protections to the order confirmation page) resulted in some friction. 

One thing I'm noticing (with the 'classic'/shortcode-based flow) is that, after a guest checks out, they are immediately asked to verify their email address. This shouldn't be the case, they should see the confirmation page.

**Comment by hsingyuc** [comment]
>I wanted to take a bit of time to test some scenarios, because the rollout of the original change (adding protections to the order confirmation page) resulted in some friction.

Sure.  I'm just trying to make the store API meet the Core expectations and expose the function so we can use it.  Let me know how I can help.

**Comment by barryhughes** [comment]
Definitely. We'd need to start by addressing the issue I just mentioned, we can't merge as-is, or we'd disrupt a large number of sites.

**Comment by hsingyuc** [comment]
>One thing I'm noticing (with the 'classic'/shortcode-based flow) is that, after a guest checks out, they are immediately asked to verify their email address. This shouldn't be the case, they should see the confirmation page.

Thank you for pointing that out!  Fixed in [4ee6572](https://github.com/woocommerce/woocommerce/pull/43834/commits/4ee657246be3d14ffc04068b5b99fa5d83003330)

**Review by barryhughes — APPROVED** [review]

**Comment by barryhughes** [comment]
Thanks! Closing and re-opening to re-start CI checks (...again 🙃).

**Comment by hsingyuc** [comment]
@barryhughes Thank you for the review!  I'm not authorized to merge the PR, could you merge it for me? Thanks!

**Pull request merged**
